### DISCUSSION TITLE
YOLOv9  model export as tflite  model

### DISCUSSION THREAD

**Discussion by HumairaBano** [ultralytics/ultralytics#10572] [Q&A]
I am trying to export yolov9 model as tflite model but got error.  Please guide me(I am using colab )

# Load the YOLOv9 model
model = YOLO('/content/runs/detect/train2/weights/best.pt')
# Export the model to TFLite format
model.export(format='tflite') # creates '.tflite' 

and error 
ONNX: export success ✅ 6.0s, saved as '/content/runs/detect/train2/weights/best.onnx' (96.9 MB)
TensorFlow SavedModel: starting TFLite export with onnx2tf 1.17.5...
TensorFlow SavedModel: export failure ❌ 26.7s: generic_type: cannot initialize type "StatusCode": an object with that name is already defined
---------------------------------------------------------------------------
ImportError                               Traceback (most recent call last)
[<ipython-input-15-9935045cba39>](https://localhost:8080/#) in <cell line: 4>()
      2 model = YOLO('/content/runs/detect/train2/weights/best.pt')
      3 # Export the model to TFLite format
----> 4 model.export(format='tflite') # creates 'yolov8n_float32.tflite'

11 frames
[/usr/local/lib/python3.10/dist-packages/tensorflow_lite_support/python/task/vision/core/tensor_image.py](https://localhost:8080/#) in <module>
     17 
     18 from tensorflow_lite_support.python.task.vision.core import color_space_type
---> 19 from tensorflow_lite_support.python.task.vision.core.pybinds import image_utils
     20 
     21 

ImportError: generic_type: cannot initialize type "StatusCode": an object with that name is already defined

**Comment by pderrenger** [comment]
Hello! It seems like you're facing a conflict issue during the TFLite export process, often due to package version mismatches.

A quick workaround to try in Colab is to restart the runtime and ensure you're using compatible versions of the packages involved in the conversion process (`tensorflow`, `onnx`, `onnx-tf`). Before exporting, you can attempt to explicitly upgrade or install specific versions of these packages and then try the export step again.

For instance:
```python
!pip install tensorflow==2.6.0 onnx onnx-tf
```

After installing or upgrading, make sure to restart the Colab runtime:
```
Runtime -> Restart runtime
```

Then, attempt the export process again:
```python
from ultralytics import YOLO

# Load the YOLOv9 model
model = YOLO('/content/runs/detect/train2/weights/best.pt')

# Export the model to TFLite format
model.export(format='tflite')
```

If issues persist, checking the specific versions of the packages involved for known compatibility issues could provide further insights. 🛠️

**Comment by glenn-jocher** [comment]
Hello! It seems like you're facing a conflict issue during the TFLite export process, often due to package version mismatches, especially when working on Colab which dynamically updates packages. A quick fix could be to refresh your Colab environment packages. Try the following steps:

1. Restart your Colab runtime to reset the environment: `Runtime > Restart runtime`.
2. Install a specific version of TensorFlow that's compatible with your needs, for example: `!pip install tensorflow==2.6.0`.
3. Ensure you have the Ultralytics YOLO package updated: `!pip install -U ultralytics`.
4. Re-run your model loading and exporting code.

```python
from ultralytics import YOLO

# Load your YOLOv9 model
model = YOLO('/content/runs/detect/train2/weights/best.pt')

# Attempt to export again
model.export(format='tflite')
```

If issues persist, it could be helpful to share the exact version of TensorFlow installed and any additional export settings you might be using. Keeping a close eye on package versions can be critical in these situations. 🛠️

Let us know how it goes!

**Comment by Varun3713-creator** [comment]
hai @pderrenger ,
which tensorflow version is compatible with yolov9

implementation ("org.tensorflow:tensorflow-lite:2.14.0")
    implementation ("org.tensorflow:tensorflow-lite-gpu:2.14.0")
    implementation ("org.tensorflow:tensorflow-lite-support:0.4.4")
    implementation ("org.tensorflow:tensorflow-lite-metadata:0.4.4")

I tried these but i am getting model load error.then i tried

implementation ("org.tensorflow:tensorflow-lite:2.16.2")
    implementation ("org.tensorflow:tensorflow-lite-gpu:2.16.2")
    implementation ("org.tensorflow:tensorflow-lite-support:0.4.4")
    implementation ("org.tensorflow:tensorflow-lite-metadata:0.4.4")

Caused by: org.gradle.api.internal.artifacts.ivyservice.DefaultLenientConfiguration$ArtifactResolveException: Could not resolve all files for configuration ':app:debugRuntimeClasspath'.
which version can i use

**Reply by pderrenger** [reply]
Hello @Varun3713-creator,

Thank you for reaching out! It looks like you're encountering issues with exporting your YOLOv9 model to TFLite format and loading it with TensorFlow Lite. Let's address your concerns step-by-step.

### Exporting YOLOv9 to TFLite

To export a YOLOv9 model to TFLite, you can use the following code snippet:

```python
from ultralytics import YOLO

# Load the YOLOv9 model
model = YOLO('/content/runs/detect/train2/weights/best.pt')

# Export the model to TFLite format
model.export(format='tflite')  # creates 'best.tflite'
```

If you encounter errors during the export process, please ensure you are using the latest versions of the Ultralytics and TensorFlow packages. You can update them using:

```bash
pip install --upgrade ultralytics tensorflow
```

### TensorFlow Lite Compatibility

Regarding the TensorFlow Lite versions, the following dependencies should generally work well:

```gradle
implementation ("org.tensorflow:tensorflow-lite:2.16.2")
implementation ("org.tensorflow:tensorflow-lite-gpu:2.16.2")
implementation ("org.tensorflow:tensorflow-lite-support:0.4.4")
implementation ("org.tensorflow:tensorflow-lite-metadata:0.4.4")
```

However, compatibility issues can arise due to various reasons. If you are facing model load errors, please ensure that:

1. **All dependencies are compatible**: Sometimes, different versions of TensorFlow Lite components may not be fully compatible with each other. It's often best to use the same version for all TensorFlow Lite components.

2. **Model Export**: Ensure that the model export process completes successfully without errors. If there are any issues during export, they might propagate to the loading phase.

3. **Environment**: Verify that your development environment is correctly set up and that there are no conflicting packages.

### Example Code for Loading TFLite Model

Here is an example of how you can load and run inference with a TFLite model in Python:

```python
import tensorflow as tf

# Load the TFLite model
interpreter = tf.lite.Interpreter(model_path="best.tflite")
interpreter.allocate_tensors()

# Get input and output tensors
input_details = interpreter.get_input_details()
output_details = interpreter.get_output_details()

# Prepare your input data
input_data = ...  # Your input data here

# Set the tensor to point to the input data to be inferred
interpreter.set_tensor(input_details[0]['index'], input_data)

# Run the inference
interpreter.invoke()

# Get the results
output_data = interpreter.get_tensor(output_details[0]['index'])
print(output_data)
```

### Further Assistance

If the issue persists, please provide a minimum reproducible example of your code and the exact error messages you are encountering. This will help us diagnose the problem more effectively. You can refer to our [Minimum Reproducible Example Guide](https://docs.ultralytics.com/help/minimum_reproducible_example) for more details.

For additional resources, you can visit our [TFLite Integration Guide](https://docs.ultralytics.com/integrations/tflite/).

I hope this helps! If you have any further questions, feel free to ask. 😊

**Comment by fogmaze** [comment]
As someone who has struggled with versioning issues, I strongly recommend using Docker.

**Reply by glenn-jocher** [reply]
@fogmaze the TFLite export error you're encountering appears related to dependency version conflicts. Let's resolve this by first ensuring you have compatible package versions in Colab:

```python
!pip uninstall -y protobuf tensorflow tflite-support
!pip install "tflite-support>=0.4.3,<=0.4.3" "protobuf<3.20"
!pip install tensorflow
```

After running these commands, try exporting again with `model.export(format='tflite')`. This should resolve the StatusCode conflict. For full TFLite export guidance, see the Ultralytics TFLite documentation.

**Comment by sagargwal** [comment]
can you please prove a docker command to export pt model in tflite format, although thier was one which was working but i am unable to find it

**Reply by sagargwal** [reply]
sorry provide

**Reply by glenn-jocher** [reply]
Here’s a ready-to-run Docker one‑liner to export a `.pt` to TFLite (mounts your current folder so the `.tflite` lands back on host); replace `best.pt` with your model path:

```bash
docker run --rm -it --gpus all -v "$PWD":/workspace -w /workspace ultralytics/ultralytics:latest \
yolo export model=best.pt format=tflite
```

CPU-only:

```bash
docker run --rm -it -v "$PWD":/workspace -w /workspace ultralytics/ultralytics:latest \
yolo export model=best.pt format=tflite device=cpu
```

INT8 example (remember to point `data` to a real dataset for calibration):

```bash
docker run --rm -it -v "$PWD":/workspace -w /workspace ultralytics/ultralytics:latest \
yolo export model=best.pt format=tflite int8=True data=path/to/data.yaml fraction=0.3
```

For available args see the [exporter reference](https://docs.ultralytics.com/reference/engine/exporter/); if you need Coral, export with `format=edgetpu` as shown in the [Edge TPU guide](https://docs.ultralytics.com/integrations/edge-tpu/).
